rr_iarrc/src/finish_line_watcher/finish_line_watcher.cpp

```diff
@@ -78,23 +84,113 @@ void ImageCB(const sensor_msgs::ImageConstPtr& msg) {
+        cv::Scalar color(0, 0, 255);
+
+        if (cv::contourArea(contours.at(i)) > min_contour_area) {
+            filteredContourPoints.insert(std::end(filteredContourPoints), std::begin(contours.at(i)),
+                                         std::end(contours.at(i)));
+            color = cv::Scalar(0, 255, 0);
+        }
+
+        // Debug draw
+        cv::drawContours(drawing, contours, i, color, 2, 8);
+    }
+
+    bool detected = false;
+
+    if (filteredContourPoints.size() > 0) {
+        // Fit a rect to the points and detect based on angle and width
+        cv::RotatedRect fitRect = cv::minAreaRect(filteredContourPoints);
+        Size2f size = fitRect.size;
+        float angle = std::abs(std::abs(fitRect.angle) - 90);
+        // Check the detection criteria (must look at both width and height because OpenCV is inconsistent about how it
+        // assigns them)
+        detected = angle < angle_cutoff && (size.width > width_cutoff || size.height > width_cutoff);
+
+        // Draw debug info
+        std::string angle_str = std::string("Angle: ") + std::to_string(angle);
+        cv::putText(drawing, angle_str, cv::Point(5, 100), cv::FONT_HERSHEY_SIMPLEX, 2, cv::Scalar(0, 143, 143), 2);
+        std::string width_str = std::string("Width: ") + std::to_string(size.width);
+        cv::putText(drawing, width_str, cv::Point(5, 200), cv::FONT_HERSHEY_SIMPLEX, 2, cv::Scalar(0, 143, 143), 2);
+        std::string height_str = std::string("Height: ") + std::to_string(size.height);
+        cv::putText(drawing, height_str, cv::Point(5, 300), cv::FONT_HERSHEY_SIMPLEX, 2, cv::Scalar(0, 143, 143), 2);
+
+        Point2f rectPoints[4];
+        fitRect.points(rectPoints);
+
+        for (int i = 0; i < 4; i++)
+            cv::line(drawing, rectPoints[i], rectPoints[(i + 1) % 4], cv::Scalar(255, 0, 0), 2, 12);
+    }
+
+    // Quick count of pixels as a final sanity check
+    auto count = cv::countNonZero(frame);
+    detected = detected && count > count_thresh;
+
+    if (publish_when_detected) {
+        // Publish when first detected but don't keep on increasing number of crosses, so make use of states again
+        if (detected && state == LOW && cooldown == 0) {
+            number_of_crosses++;
+            ROS_INFO_STREAM("Finish line crossed: " << to_string(number_of_crosses));
+            state = HIGH;
+            cooldown = cooldown_value;
+
+        } else if (state == HIGH && !detected) {
+            state = LOW;
+        }
+
+    } else {
+        // When the line is first detected, set to a high state then when we stop detecting (we've crossed), report that
+        if (state == LOW && detected && cooldown == 0) {
+            state = HIGH;
 
-    auto width = getWidth(frame);
+        } else if (state == HIGH && !detected) {
+            state = LOW;
+            number_of_crosses++;
+            cooldown = cooldown_value;
+            ROS_INFO_STREAM("Finish line crossed: " << to_string(number_of_crosses));
```

Comment:
Maybe compress this with the previous block so that you set `incrementCrossNum = True && publish_when_detected` in one block and `incrementCrossNum = True && !publish_when_detected` in the other.
Then have an if-statement later that checks incrementCrossNum where you actually set the number of crosses, the cooldown, and prints it. 
Just so it is easier to read.

Reply:
You should try to combine the two block with the `incrementCrossNum = True && publish_when_detected` and `incrementCrossNum = True && !publish_when_detected` trick.